stdlib/public/runtime/MetadataLookup.cpp

```diff
@@ -1240,6 +1240,20 @@ _gatherGenericParameters(const ContextDescriptor *context,
     };
   };
 
+  auto generics = context->getGenericContext();
+  unsigned numOfKeyArgs = 0;
+  auto hasOnlyKeyArgs = false;
```

Comment:
No, let's not do this; it feels to brittle to count the arguments and take different code paths.

I think it's fine for the callers to provide the non-key arguments because we can still check that it satisfies the same-type requirement. If `G<T>.Inner` requires that `T == Int`, we should not allow instantiating the type `G<String>.Inner`. However, you could instead add a new entry point with the different behavior.

Reply:
We don't allow `G<String>.Inner` with this patch either. That type has no key arguments so when reading from the key arguments in the metadata and using those same values to reconstruct the same type we'll fail for `G<Int>.Inner` because the metadata says there is no generic arguments and trying to use the same generic arguments (nothing) to `_swift_instantiateCheckGenericMetadata` will fail because it expects non-key arguments only to pass nothing to the accessor.

Reply:
I'm ok putting the key argument version into a separate entry point, but it's going to do the same thing that `_gatherGenericParameters` basically if you're fine with the duplication.

Reply:
Maybe you want to wrap `_checkGenericRequirements()` instead?

Reply:
I'd still like to have things like share classes get filled out for me, but it does look like `_checkGenericRequirements` fills in witness tables.

Reply:
I made it so this instantiate function no longer wraps _gatherGenericParameters and instead just wraps _checkGenericRequirements which means I have to provide shape classes now, but it's not too difficult.